File: Java/P3/irrgaten/src/irrgarten/Player.java

package irrgarten;
import java.util.ArrayList;

/**
 * @brief   Funcionalidades de los jugadores
 * @author  Isabel Morro Tabares
 */
public class Player {
   
    // Declaración de atributos
    
    private static final int MAX_WEAPONS = 2;
    private static final int MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH = 10;
    private static final int HITS2LOSE = 3;
    
    private String name;
    private char number;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;
    private int consecutiveHits;
    
    private ArrayList<Weapon> weapons;
    private ArrayList<Shield> shields;
    
    // Constructor
    
    public Player (char number, float intelligence, float strength){
        this.number = number;
        this.name = "Player #" + number;
        this.intelligence = intelligence;
        this.strength = strength;
        
        weapons = new ArrayList<>();
        shields = new ArrayList<>();
        
        health = INITIAL_HEALTH;
        consecutiveHits = 0;
        
        row = -1;
        col = -1;
    }
    
    // Métodos privados
    
    private void receiveWeapon(Weapon w){
        
        boolean discard;
        
        System.out.print(weapons.size());
        
        for (Weapon wi : weapons){
            discard = wi.discard();
            if(discard)
                weapons.remove(wi);
        }
        
        int size = weapons.size();
        
        if (size < MAX_WEAPONS) weapons.add(w);
        
    }
    
    private void receiveShield(Shield s){
        
        boolean discard;
        
        System.out.print(shields.size());
        
        for (Shield si : shields){
            discard = si.discard();
            if (discard)
                shields.remove(si);
        }
        
        int size = shields.size();
        
        if (size < MAX_SHIELDS) shields.add(s);
    }
    
    private Weapon newWeapon() {
        
        float power = Dice.weaponPower();
        int uses = Dice.usesLeft();
        Weapon w = new Weapon(power, uses);
        return w;
    }
    
    private Shield newShield(){
        
        float protection = Dice.shieldPower();
        int uses = Dice.usesLeft();
        Shield s = new Shield(protection, uses);
        return s;
    }
    
    private float sumWeapons(){
        float suma = 0f;
        for (int i=0; i < weapons.size(); i++)
            suma += weapons.get(i).attack();
        return suma;
    }
    
    private float sumShields(){
        
        float suma = 0f;
        for(int i=0; i < shields.size(); i++)
            suma += shields.get(i).protect();
        return suma;
    }
    
    private float defensiveEnergy(){
        return sumShields() + intelligence;
    }
    
    private boolean manageHit (float receivedAttack){
        
        float defense = defensiveEnergy();
        boolean lose;
        
        if (defense < receivedAttack){
            gotWounded();
            incConsecutiveHits();
        }
        else resetHits();
        
        if ( consecutiveHits == HITS2LOSE || dead()){
            resetHits();
            lose = true;
        }
        else lose = false;
        
        return lose;
    }
    
    private void resetHits(){
        consecutiveHits = 0;
    }
    
    private void gotWounded(){
        if (health > 0)
            health--;
    }
    
    private void incConsecutiveHits(){
        consecutiveHits++;
    }
    
    
    // Métodos públicos
    
    public void resurrect (){
        weapons.clear(); 
        shields.clear();
        health = INITIAL_HEALTH;
        resetHits();
    }
    
    // Getters
    
    public int getRow(){
        return row;
    }
    
    public int getCol(){
        return col;
    }
    
    public char getNumber(){
        return number;
    }
    
    public void setPos (int row, int col){
        this.row = row;
        this.col = col;
    }
    
    public boolean dead(){
        return health <= 0;
    }

    public Directions move(Directions direction, ArrayList<Directions> validMoves){
        
        int size = validMoves.size();
        boolean contained = validMoves.contains(direction);
        Directions result;
        
        if (size>0 && !contained)
            result = validMoves.get(0);
        else
            result = direction;
        
        return direction;
    }
    
    public float attack(){
        return sumWeapons() + strength;
    }
    
    public boolean defend (float receivedAttack){
        return manageHit(receivedAttack);
    }
    
    public void receivedReward(){
        
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();
        
        for (int i=0; i < wReward; i++){
            Weapon wnew = newWeapon();
            receiveWeapon(wnew);
        }
        
        for (int i=0; i < sReward; i++){
            Shield snew = newShield();
            receiveShield(snew);
        }
        
        int extraHealth = Dice.healthReward();
        health += extraHealth;
    }
    
    @Override
    public String toString(){
        String cadena = "[ Nombre: " + name + "\n"
                        + " Numero: " + number + "\n"
                        + " Inteligencia: " + intelligence + "\n"
                        + " Fuerza: " + strength + "\n"
                        + " Salud: " + health + "\n"
                        + " Posicion: (" + row + ", " + col + ") " + "\n"
                        + " Impactos consecutivos: " + consecutiveHits + "]\n";
        return cadena;
    }
   
}

File: Ruby/P3/Irrgarten/player.rb
#encoding:utf-8

=begin

    Autor: Isabel Morro Tabares 

    Funcionalidades de los jugadores

=end

require_relative 'dice'
require_relative 'directions'
require_relative 'weapon'
require_relative 'shield'

module Irrgarten
    class Player 
    
    private 

        @@MAX_WEAPONS = 2
        @@MAX_SHIELDS = 3
        @@INITIAL_HEALTH = 10
        @@HITS_2_LOSE = 3

        def initialize (number, intelligence, strength)
            @number = number
            @name = "Player #" + @number
            @intelligence = intelligence
            @strength = strength
            @health = @@INITIAL_HEALTH
            @row = 0
            @col = 0
            @consecutive_hits = 0

            @weapons = Array.new
            @shields = Array.new

        end

        # Métodos privados

        def receive_weapon(w)


            @weapons.each do |wi|
                discard = wi.discard()
                if (discard)
                    @weapons.delete(wi)
                end
            end

            size = @weapons.size

            if (size < @@MAX_WEAPONS)
                @weapons << w  
            end

        end

        def receive_shields(s)
        
            @shields.each do |si|
                discard = si.discard()
                if (discard)
                    @shields.delete(si)
                end
            end

            size = @shields.size

            if (size < @@MAX_SHIELDS)
                @shields << s 
            end
        
        end

        def new_weapon
            power = Dice.weapon_power
            uses = Dice.uses_left
            w = Wepon.new(power, uses)
        end

        def new_shield            
            protection = Dice.shield_power
            uses = Dice.uses_Left
            s = Shield.new(protection, uses)
        end

        def sum_weapons
            suma = 0
            for i in (0..@weapons.length)
                suma = suma + @weapons[i].attack
            end
        end

        def sum_shields
            suma = 0
            for i in (0..@shields.length)
                suma = suma + @shields[i].protect
            end
        end

        def defensive_energy
            sum_shields() + intelligence
        end

        def manage_Hit(receive_attack)

            defense = defensive_energy()

            if (defense < receiv_aAttack)
                got_wounded()
                inc_consecutive_hits()
            else
                reset_hits()
            end

            if (consecutive_hits == @@HITS_2_LOSE || dead())
                reset_hits()
                lose = true
            else
                lose = false 
            end

            return lose

        end

        def reset_hits
            consecutive_hits = 0
        end

        def got_wounded
            if health > 0
                health = health - 1
            end
        end

        def inc_consecutive_hits
            consecutive_hits += 1
        end

    public 

        # Métodos públicos

        def resurrect
            @weapons.clear 
            @shields.clear 
            @health = @@INITIAL_HEALTH
            reset_hits()
        end

        def row
            @row
        end

        def col
            @col
        end

        def number
            @number
        end

        def pos (row, col)
            @row = row
            @col = col
        end

        def dead
            (@health <= 0)
        end

        def move (direction, valid_moves)

            size = valid_moves.length
            contained = valid_moves.include?(direction)

            if (size > 0 && !contained)
                first_element = valid_moves[0]
                resultado = first_element
            else
                resultado = direction
            end

            return resultado

        end

        def attack
            sum_weapons() + strength
        end

        def defend
            manage_hit(receive_attack)
        end

        def receive_reward

            w_reward = Dice.weapons_reward()
            s_reward = Dice.shields_reward()

            for i in (0..w_reward)
                w_new = new_weapon()
                receive_weapon(w_new)
            end

            for i in (0..s_reward)
                s_new = new_shield()
                receive_rsield(s_new)
            end

            extra_health = Dice.health_reward()
            health += extra_health

        end

        # To string
        def to_s
            cadena = ""
            cadena  += "[ Nombre: " + @name + "\n"
            cadena  += " Numero: " + @number + "\n"
            cadena  += " Inteligencia: " + @intelligence.to_s + "\n"
            cadena  += " Fuerza: " + @strength.to_s + "\n"
            cadena  += " Salud: " + @health.to_s + "\n"
            cadena  += " Posicion: (" << @row.to_s + ", " + @col.to_s + ") " + "\n"
            cadena  += " Impactos consecutivos: " + @consecutive_hits.to_s + "]\n" 
        end

    
    end
end
